pallets/token-dealer/src/lib.rs

```diff
@@ -14,157 +14,350 @@
-
-#[derive(Encode, Decode)]
-pub enum XCMPMessage<XAccountId, XBalance> {
-	/// Transfer tokens to the given account from the Parachain account.
-	TransferToken(XAccountId, XBalance),
-}
+pub mod support;
 
-type BalanceOf<T> =
-	<<T as Trait>::Currency as Currency<<T as frame_system::Trait>::AccountId>>::Balance;
+use frame_support::pallet;
 
-type RelayAccountId = sp_runtime::AccountId32;
+pub use pallet::*;
 
-/// Configuration trait of this pallet.
-pub trait Trait: frame_system::Trait {
-	/// Event type used by the runtime.
-	type Event: From<Event<Self>> + Into<<Self as frame_system::Trait>::Event>;
+#[pallet]
+pub mod pallet {
+	use cumulus_primitives::{relay_chain::Balance as RelayChainBalance, ParaId};
+	use frame_support::{pallet_prelude::*, traits::Get};
+	use frame_system::pallet_prelude::*;
+	use sp_runtime::traits::{AtLeast32BitUnsigned, Convert};
+	use sp_std::prelude::*;
+	use xcm::v0::{
+		Error as XcmError, ExecuteXcm, Junction, MultiAsset, MultiLocation, NetworkId, Order, Xcm,
+	};
+	use xcm_executor::traits::LocationConversion;
 
-	/// The sender of upward messages.
-	type UpwardMessageSender: UpwardMessageSender<Self::UpwardMessage>;
+	#[derive(Encode, Decode, Eq, PartialEq, Clone, Copy, RuntimeDebug)]
+	/// Identity of chain.
+	pub enum ChainId {
+		/// The relay chain.
+		RelayChain,
+		/// A parachain.
+		ParaChain(ParaId),
+	}
 
-	/// The upward message type used by the Parachain runtime.
-	type UpwardMessage: codec::Codec + BalancesMessage<RelayAccountId, BalanceOf<Self>>;
+	#[derive(Encode, Decode, Eq, PartialEq, Clone, RuntimeDebug)]
+	/// Identity of cross chain currency.
+	pub struct XCurrencyId {
+		/// The reserve chain of the currency. For instance, the reserve chain
+		/// of DOT is Polkadot.
+		pub chain_id: ChainId,
+		/// The identity of the currency.
+		pub currency_id: Vec<u8>,
+	}
 
-	/// Currency of the runtime.
-	type Currency: Currency<Self::AccountId>;
+	impl Into<MultiLocation> for XCurrencyId {
+		fn into(self) -> MultiLocation {
+			MultiLocation::X1(Junction::GeneralKey(self.currency_id))
+		}
+	}
 
-	/// The sender of XCMP messages.
-	type XCMPMessageSender: XCMPMessageSender<XCMPMessage<Self::AccountId, BalanceOf<Self>>>;
-}
+	/// Pallet for executing cross-chain transfers
+	#[pallet::pallet]
+	pub struct Pallet<T>(PhantomData<T>);
 
-decl_event! {
-	pub enum Event<T> where
-		AccountId = <T as frame_system::Trait>::AccountId,
-		Balance = BalanceOf<T>
-	{
-		/// Transferred tokens to the account on the relay chain.
-		TransferredTokensToRelayChain(RelayAccountId, Balance),
-		/// Transferred tokens to the account on request from the relay chain.
-		TransferredTokensFromRelayChain(AccountId, Balance),
-		/// Transferred tokens to the account from the given parachain account.
-		TransferredTokensViaXCMP(ParaId, AccountId, Balance, DispatchResult),
-	}
-}
+	/// The shape of AccountId for (most) substrate chains (not Moonbeam, which is H160 so 20 bytes)
+	type SubstrateAccountId = [u8; 32];
```

Comment:
does it make sense to `use sp_runtime::AccountId32` for this?

Reply:
https://substrate.dev/rustdocs/v3.0.0/sp_runtime/struct.AccountId32.html

I don't know, what would the benefit be? The `id` field of the cross-chain message with format `MultiLocation::X1(Junction::AccountId32` is `[u8; 32]` and although `sp_runtime::AccountId32` can `.into()` `[u8; 32`, are there benefits here to use that type and add `.into()` for every time we use it to specify the destination as a cross-chain message. 

Related question: Does it possibly simplify the interface for polkadot-js?  i.e. Is `[u8; 32]` less supported in polkadot-js than `sp_runtime::AccountId32`?

Reply:
> The id field of the cross-chain message with format MultiLocation::X1(Junction::AccountId32 is [u8; 32]

Ahh, I see! I think this is what I was wondering about with the account id converter and stuff. Shouldn't messages addressed to Moonbeam use `Junction::AccountId20`? Is that a thing?

Reply:
I think the idea is that messages sent to moonbeam should be addressed in a moonbeam-friendly way. Just like when I mail a letter to Japan I better address it in a way the Japanese postal service will understand it so it gets to the right spot.

I admit I have not fully understood all the code yet, so I could be wrong.

Reply:
Your intuition is correct, but `SubstrateAccountId` is only used when the RECIPIENT is a chain that has a 32 byte account id. 

For messages when our chain is the recipient, they are processed by the `cumulus_parachain_system` pallet's `DownwardMessageHandler` associated type. This type is set equal to `XcmHandler` in the runtime and, in particular, it uses the logic in the `TransactAsset` implementation of `MultiCurrencyAdapter` (in `token-dealer/src/support.rs`) to execute messages in the downward message queue.